### PR TITLE
Updates k8s ParseNode() to Handle Multiple IPs

### LINKED ISSUE #20787: Better handle when K8s Nodes have more than one IP address per address type in their status field
https://github.com/kubernetes/kubernetes/issues/61921 indicates that K8s assumes that a Node has one IP per type, as seen in https://github.com/kubernetes/kubernetes/blob/8f3b2813dcd6e3f2c9b260bedaeddbdd0ea5c0cc/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L711 and https://github.com/kubernetes/kubernetes/blob/master/pkg/util/node/node.go#L81.

There are two approaches:

1) Add a warning log to the user to say that we detected multiple IPs per type and this may impact the ipcache by having potential conflicts with pod IPs. As seen in https://github.com/cilium/cilium/issues/20243.
2) Enforce the 1 IP per type and reject all other IPs of the same type. Use the same preference [logic](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/node/node.go#L81) as K8s.

Follow up from https://github.com/cilium/cilium/issues/20243.

### PR DESCRIPTION
**Pull request by danehans**
This PR updates `ParseNode()` in the k8s pkg to emit a warning-level log message when a k8s node contains multiple IPs of the same address type and family.

Fixes: https://github.com/cilium/cilium/issues/20787
Supercedes: https://github.com/cilium/cilium/pull/20811

```release-note
When a k8s node contains multiple addresses of the same type and family, Cilium will now emit a warning-level log message stating: "Detected multiple IPs of the same address type, Cilium will only consider the first IP in the Node resource"
```

### CODE REVIEW

**Comment by danehans** [comment]
cc: @christarazi since you reviewed #20811 and assigned me #20787.

File: pkg/k8s/node.go
Comment (asauber) [inline pkg/k8s/node.go]: This logic doesn't seem sufficiently precise to me. Could it be the case that we have exactly two v4 CIDRs and we should still print the warning?

File: pkg/k8s/node_test.go
Comment (asauber) [inline pkg/k8s/node_test.go:101]: We test the not-detected case here. Is there a way that we can test the detected case?

File: pkg/k8s/node.go
Comment (danehans) [reply pkg/k8s/node.go]: @asauber thanks for the review. I have updated `ParseNode()` to now begin the next iteration of node addresses if the address is invalid. A node can have at most two "InternalIP" and "ExternalIP" addresses as long as each is a separate address family, e.g. IPv4 or IPv6. See https://github.com/cilium/cilium/pull/20811#issuecomment-1235122629 for additional context. I have also removed `len(k8sNode.Spec.PodCIDRs) != 2` from this LOC since it's being checked later in `ParseNode()` and is not applicable to the management of node status addresses.

File: pkg/k8s/node_test.go
Comment (danehans) [reply pkg/k8s/node_test.go:101]: I added a unit test where the node contains multiple InternalIP/ExternalIP/IPv4/IPv6 addresses, PTAL. The newly added test created a node with 8 `NodeAddress`es, two of each type/family, and asserts that only 4 should be added to the node:
```sh
START: node_test.go:20: K8sSuite.TestParseNode
level=warning msg="Detected multiple IPs of the same address type, Cilium will only consider the first IP in the Node resource" k8sNodeID= node=node2 nodeName=node2 subsys=k8s type=InternalIP
level=warning msg="Detected multiple IPs of the same address type, Cilium will only consider the first IP in the Node resource" k8sNodeID= node=node2 nodeName=node2 subsys=k8s type=InternalIP
level=warning msg="Detected multiple IPs of the same address type, Cilium will only consider the first IP in the Node resource" k8sNodeID= node=node2 nodeName=node2 subsys=k8s type=ExternalIP
level=warning msg="Detected multiple IPs of the same address type, Cilium will only consider the first IP in the Node resource" k8sNodeID= node=node2 nodeName=node2 subsys=k8s type=ExternalIP
PASS: node_test.go:20: K8sSuite.TestParseNode	0.006s
```

**Review by asauber — APPROVED** [review]
LGTM. Thanks for the updates.

**Review by tommyp1ckles — APPROVED** [review]

**Review by christarazi — COMMENTED** [review]
Thanks for the PR!

Could you squash the two commits into one? No need for two commits that are incomplete without each other. I realize that the first commit is from a different author so please use `Co-authored-by: ...` to properly attribute yourself and the original author in the squashed commit.

**Comment by christarazi** [comment]
@danehans Just a heads up that pushing to the PR resets the CI state so we'd need to run it again. It also passed due to the presence of the `ready-to-merge` label that was set by MLH.

If you pushed to rebase onto latest main, then it's not necessary unless there's a conflict that needs to be resolved. We merge PRs with a rebase and merge strategy, so the PR does not need to be fully synced with main before merging.

**Comment by danehans** [comment]
Commit 10903f0 rebases due to merge conflict.

**Comment by danehans** [comment]
Commit 59b917f  was failing `checkpatch` b/c the commit message was too long.

**Comment by danehans** [comment]
Commit 91bc601 was failing `checkpatch` b/c the co-authored-by was incorrectly formatted. `checkpatch` is now passing locally.

**Comment by danehans** [comment]
Commit 91bc601 was failing `checkpatch` b/c the co-authored-by was incorrectly formatted. `checkpatch` is now passing locally.

**Comment by danehans** [comment]
@christarazi not sure why this PR is causing the Ingress/Gateway API conformance tests to fail. Could it be a test flake?

**Comment by danehans** [comment]
@christarazi not sure if the Ingress/Gateway API conformance tests are required and why they would be failing due to this PR.

**Comment by christarazi** [comment]
Rebased to re-run tests. We had some CI outage over the weekend so better to start fresh. I created a tracking [issue](https://github.com/cilium/cilium/issues/25595) for the ConformanceIngress flake failure.

**Comment by danehans** [comment]
The `runtime` job is failing due to:
```
	 [18:08:09]     adverts_test.go:146: 
	 [18:08:09]         	Error Trace:	/home/vagrant/go/src/github.com/cilium/cilium/pkg/bgpv1/test/adverts_test.go:146
	 [18:08:09]         	Error:      	Received unexpected error:
	 [18:08:09]         	            	did not receive expected peering state ["ESTABLISHED"], context deadline exceeded
	 [18:08:09]         	Test:       	Test_PodCIDRAdvert
	 [18:08:09] time="2023-05-23T18:08:09Z" level=info msg="GoBGP test instance: stopping"
	 [18:08:09] time="2023-05-23T18:08:09Z" level=info msg="Delete a peer configuration" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGP subsys=basic
	 [18:08:09] time="2023-05-23T18:08:09Z" level=debug msg="stop connect loop" Key=172.16.100.1 Topic=Peer asn=65011 component=tests.BGP subsys=basic
	 [18:08:09] time="2023-05-23T18:08:09Z" level=debug msg="freed fsm.h" Key=172.16.100.1 State=BGP_FSM_ACTIVE Topic=Peer asn=65011 component=tests.BGP subsys=basic
	 [18:08:09] level=info msg=Stopping subsys=hive
	 [18:08:09] time="2023-05-23T18:08:09Z" level=info msg="deleting dummy links"
	 [18:08:09] === RUN   Test
	 [18:08:09] level=warning msg="Failed to downgrade endpoint to original ENI datapath. Previous datapath is still intact and endpoint connectivity is not affected." error="failed to create new rule: unable to add new rule: fake error" rule="ip rule 111: from 172.16.1.0/24 to all table 11" subsys=linux-routing
	 [18:08:09] --- FAIL: Test_PodCIDRAdvert (15.14s)
```

Rebasing 7e150e3  since the `bgpv1/test` pkg is not part of this PR's commit history.

**Comment by christarazi** [comment]
/test-1.26-net-next

**Comment by christarazi** [comment]
/test-1.26-net-next

**Comment by danehans** [comment]
The `k8s-1.26-kernel-net-next` job is failing due to the agent health check failing:
```
05:30:35 STEP: Cilium is not ready yet: connectivity health is failing: Cluster connectivity is unhealthy on 'cilium-4r2kv': Exitcode: 1 
Err: exit status 1
Stdout:
 	 
Stderr:
 	 Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
	 Error: Cannot get status/probe: Put "[http://%2Fvar%2Frun%2Fcilium%2Fhealth.sock/v1beta/status/probe":](http://%2Fvar%2Frun%2Fcilium%2Fhealth.sock/v1beta/status/probe) dial unix /var/run/cilium/health.sock: connect: no such file or directory
	 
	 command terminated with exit code 1
````

**Comment by danehans** [comment]
@christarazi it appears that https://github.com/cilium/cilium/pull/25304#issuecomment-1561375004 did not create a GH issue. Does this command need to be executed by a maintainer?

**Comment by christarazi** [comment]
/test-1.26-net-next

**Pull request merged**
